temporalio/client.py

```diff
@@ -0,0 +1,935 @@
+                workflow ID.
+            follow_runs: If true (default), workflow runs will be continually
+                fetched, until the most recent one is found. If false, the first
+                result is used.
+
+        Returns:
+            Result of the workflow after being converted by the data converter.
+
+        Raises:
+            Exception: Any failure of the workflow.
+        """
+        if starting_run_id == WorkflowHandle.SELF_RUN_ID:
+            starting_run_id = self._run_id
+        req = temporalio.api.workflowservice.v1.GetWorkflowExecutionHistoryRequest(
+            namespace=self._client.namespace,
+            execution=temporalio.api.common.v1.WorkflowExecution(
+                workflow_id=self._id, run_id=starting_run_id or ""
+            ),
+            wait_new_event=True,
+            history_event_filter_type=temporalio.api.enums.v1.HistoryEventFilterType.HISTORY_EVENT_FILTER_TYPE_CLOSE_EVENT,
+            skip_archival=True,
+        )
+        while True:
+            resp = await self._client.service.get_workflow_execution_history(req)
+            # Continually ask for pages until we get close
+            if len(resp.history.events) == 0:
+                req.next_page_token = resp.next_page_token
+                continue
+            elif len(resp.history.events) != 1:
+                raise RuntimeError(
+                    f"Expected single close event, got {len(resp.history.events)}"
+                )
+            event = resp.history.events[0]
+            if event.HasField("workflow_execution_completed_event_attributes"):
+                complete_attr = event.workflow_execution_completed_event_attributes
+                # Follow execution
+                if follow_runs and complete_attr.new_execution_run_id:
+                    req.execution.run_id = complete_attr.new_execution_run_id
+                    req.next_page_token = b""
+                    continue
+                # Ignoring anything after the first response like TypeScript
+                if not complete_attr.result:
+                    return cast(T, None)
+                results = await self._client.data_converter.decode(
+                    complete_attr.result.payloads
+                )
+                if not results:
+                    return cast(T, None)
+                elif len(results) > 1:
+                    logger.warning("Expected single result, got %s", len(results))
+                return cast(T, results[0])
+            elif event.HasField("workflow_execution_failed_event_attributes"):
+                fail_attr = event.workflow_execution_failed_event_attributes
+                # Follow execution
+                if follow_runs and fail_attr.new_execution_run_id:
+                    req.execution.run_id = fail_attr.new_execution_run_id
+                    req.next_page_token = b""
+                    continue
+                raise WorkflowFailureError(
```

Comment:
I'd consider raising `WorkflowFailureError` from `TemporalFailure` to integrate with python's built-in exception chaining.

Reply:
From `TemporalFailure`? Can you clarify? I do set `__cause__` which is what `from` does. Are you saying I should:

```python
cause = await temporalio.failure.FailureError.from_proto(fail_attr.failure, self._client.data_converter)
raise WorkflowFailureError() from cause
```

Why? That just removes my typing.